### ISSUE TITLE
UnsupportedCommandException: sendKeysToActiveElement when trying to submit form via 'Enter'

### ISSUE THREAD

**Issue by dude8** [SeleniumHQ/selenium#4546]
## Meta -
OS:  
Ubuntu 14.04
Selenium Version:  
3.4.0 standalone
Browser:  
Firefox
Browser Version:  
54 (64 bit)
55 (64 bit)
Gecko Driver
0.16.1

## Expected Behavior -
$this->keys() can be used to submit a form

## Actual Behavior -
`org.openqa.selenium.UnsupportedCommandException: sendKeysToActiveElement` is thrown and test fails. The Test works fine in Chrome Browser

## Steps to reproduce -
The error can be reproduced with the following code snippet.

```php
<?php
class KeysNotWorkingTest extends PHPUnit_Extensions_Selenium2TestCase
{
    protected function setUp() {
        $this->setupSpecificBrowser(array(
            "browserName" => "firefox",
            "sessionStrategy" => "shared",
            "host" => "localhost",
            "port" => 4444
        ));
        $this->setBrowserUrl("http://www.google.com");
    }
    public function testKeysNotWorking() {
        $this->url('/');
        $this->byId('lst-ib')->value('selenium');
        //PHPUnit_Extensions_Selenium2TestCase_Keys::RETURN_ doesn't work either
        $this->keys(PHPUnit_Extensions_Selenium2TestCase_Keys::ENTER);
    }
}
```
Selenium Server Log:
```
10:01:26.507 INFO - Executing: [new session: Capabilities [{browserName=firefox}]])
10:01:26.510 INFO - Creating a new session for Capabilities [{browserName=firefox}]
1503475286548	geckodriver	INFO	Listening on 127.0.0.1:11163
1503475286632	geckodriver::marionette	INFO	Starting browser /usr/lib/firefox/firefox with args ["-marionette"]
1503475286992	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503475286995	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503475287000	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.
1503475287001	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.
1503475287172	Marionette	INFO	Enabled via --marionette

(firefox:18385): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(firefox:18385): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed

(firefox:18385): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(firefox:18385): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed
1503475287244	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503475287245	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.
1503475287249	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503475287250	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.

(/usr/lib/firefox/firefox:18453): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(/usr/lib/firefox/firefox:18453): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed

(/usr/lib/firefox/firefox:18453): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(/usr/lib/firefox/firefox:18453): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed
1503475288845	Marionette	INFO	Listening on port 33678
JavaScript error: chrome://marionette/content/server.js, line 337: NS_ERROR_SOCKET_ADDRESS_IN_USE: Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]
1503475289020	Marionette	DEBUG	loaded listener.js
10:01:29.079 INFO - Detected dialect: W3C
10:01:29.081 INFO - Done: [new session: Capabilities [{browserName=firefox}]]
10:01:29.084 INFO - Executing: [get current window handle])
10:01:29.116 INFO - Done: [get current window handle]
10:01:29.118 INFO - Executing: [get: http://www.google.com/])
1503475289132	Marionette	DEBUG	Received DOM event "beforeunload" for "about:blank"
1503475289842	Marionette	DEBUG	Received DOM event "pagehide" for "about:blank"
1503475289842	Marionette	DEBUG	Received DOM event "unload" for "about:blank"
1503475290190	Marionette	DEBUG	Received DOM event "DOMContentLoaded" for "https://www.google.at/?gfe_rd=cr&ei=VzadWdLiEI7KXtqYgHg&gws_rd=ssl"
1503475291629	Marionette	DEBUG	Received DOM event "pageshow" for "https://www.google.at/?gfe_rd=cr&ei=VzadWdLiEI7KXtqYgHg&gws_rd=ssl"
10:01:31.670 INFO - Done: [get: http://www.google.com/]
10:01:31.673 INFO - Executing: [find element: By.id: lst-ib])
10:01:31.711 INFO - Done: [find element: By.id: lst-ib]
10:01:31.715 INFO - Executing: [send keys: 0 [[FirefoxDriver: firefox on ANY (f02334aa-8a95-4558-9173-672b60c82ceb)] -> id: lst-ib], [s, e, l, e, n, i, u, m]])
10:01:31.931 INFO - Done: [send keys: 0 [[FirefoxDriver: firefox on ANY (f02334aa-8a95-4558-9173-672b60c82ceb)] -> id: lst-ib], [s, e, l, e, n, i, u, m]]
10:01:31.932 INFO - Executing: [send keys to active: []])
10:01:31.934 WARN - Exception thrown
org.openqa.selenium.UnsupportedCommandException: sendKeysToActiveElement
Build info: version: '3.4.0', revision: 'unknown', time: 'unknown'
System info: host: 'ecDev14', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-78-generic', java.version: '1.8.0_144'
Driver info: driver.version: RemoteWebDriver
	at org.openqa.selenium.remote.http.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:220)
	at org.openqa.selenium.remote.http.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:118)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:158)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:82)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:637)
	at org.openqa.selenium.remote.RemoteExecuteMethod.execute(RemoteExecuteMethod.java:35)
	at org.openqa.selenium.remote.RemoteKeyboard.sendKeys(RemoteKeyboard.java:35)
	at org.openqa.selenium.support.events.internal.EventFiringKeyboard.sendKeys(EventFiringKeyboard.java:42)
	at org.openqa.selenium.remote.server.handler.interactions.SendKeyToActiveElement.call(SendKeyToActiveElement.java:56)
	at org.openqa.selenium.remote.server.handler.interactions.SendKeyToActiveElement.call(SendKeyToActiveElement.java:31)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.openqa.selenium.remote.server.DefaultSession$1.run(DefaultSession.java:176)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
10:01:31.934 WARN - Exception: sendKeysToActiveElement
Build info: version: '3.4.0', revision: 'unknown', time: 'unknown'
System info: host: 'ecDev14', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-78-generic', java.version: '1.8.0_144'
Driver info: driver.version: RemoteWebDriver
10:01:31.941 INFO - Executing: [delete session: 40f4068c-23f8-435e-b68b-657d50ef4bac])
1503475291972	Marionette	INFO	New connections will no longer be accepted
JavaScript error: resource://shield-recipe-client/lib/RecipeRunner.jsm, line 249: TypeError: log is undefined
[Child 18453] WARNING: pipe error (18): Connection reset by peer: file /build/firefox-Zj9gRO/firefox-55.0.2+build1/ipc/chromium/src/chrome/common/ipc_channel_posix.cc, line 353
[Child 18453] WARNING: pipe error (35): Connection reset by peer: file /build/firefox-Zj9gRO/firefox-55.0.2+build1/ipc/chromium/src/chrome/common/ipc_channel_posix.cc, line 353
[Child 18453] WARNING: pipe error (3): Connection reset by peer: file /build/firefox-Zj9gRO/firefox-55.0.2+build1/ipc/chromium/src/chrome/common/ipc_channel_posix.cc, line 353
10:01:32.373 INFO - Done: [delete session: 40f4068c-23f8-435e-b68b-657d50ef4bac]

```

**Comment by barancev** [comment]
Update geckodriver to 0.18

**Comment by dude8** [comment]
Updating Geckodriver didn't help.

Log with geckodriver 0.18.0:
```
10:56:53.149 INFO - Executing: [new session: Capabilities [{browserName=firefox}]])
10:56:53.208 INFO - Creating a new session for Capabilities [{browserName=firefox}]
1503478613732	geckodriver	INFO	geckodriver 0.18.0
1503478613735	geckodriver	INFO	Listening on 127.0.0.1:17185
1503478613932	geckodriver::marionette	INFO	Starting browser /usr/lib/firefox/firefox with args ["-marionette"]
1503478614334	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503478614335	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503478614341	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.
1503478614342	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.
1503478614562	Marionette	INFO	Enabled via --marionette

(firefox:27438): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(firefox:27438): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed

(firefox:27438): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(firefox:27438): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed
1503478614670	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503478614671	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.
1503478614694	addons.xpi	WARN	Add-on langpack-en-GB@firefox.mozilla.org is not compatible with application version.
1503478614695	addons.xpi	WARN	Add-on langpack-en-ZA@firefox.mozilla.org is not compatible with application version.

(/usr/lib/firefox/firefox:27495): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(/usr/lib/firefox/firefox:27495): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed

(/usr/lib/firefox/firefox:27495): GLib-GObject-CRITICAL **: g_object_ref: assertion 'object->ref_count > 0' failed

(/usr/lib/firefox/firefox:27495): GLib-GObject-CRITICAL **: g_object_unref: assertion 'object->ref_count > 0' failed
1503478616422	Marionette	INFO	Listening on port 42853
JavaScript error: chrome://marionette/content/server.js, line 337: NS_ERROR_SOCKET_ADDRESS_IN_USE: Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]
1503478616641	Marionette	DEBUG	loaded listener.js
10:56:56.751 INFO - Detected dialect: W3C
10:56:56.812 INFO - Done: [new session: Capabilities [{browserName=firefox}]]
10:56:56.822 INFO - Executing: [get current window handle])
10:56:56.856 INFO - Done: [get current window handle]
10:56:56.863 INFO - Executing: [get: http://www.google.com/])
1503478616870	Marionette	DEBUG	Received DOM event "beforeunload" for "about:blank"
1503478617702	Marionette	DEBUG	Received DOM event "pagehide" for "about:blank"
1503478617702	Marionette	DEBUG	Received DOM event "unload" for "about:blank"
1503478618079	Marionette	DEBUG	Received DOM event "DOMContentLoaded" for "https://www.google.at/?gfe_rd=cr&ei=YUOdWaGeE4bEXrfkoegD&gws_rd=ssl"
1503478619221	Marionette	DEBUG	Received DOM event "pageshow" for "https://www.google.at/?gfe_rd=cr&ei=YUOdWaGeE4bEXrfkoegD&gws_rd=ssl"
10:56:59.246 INFO - Done: [get: http://www.google.com/]
10:56:59.252 INFO - Executing: [find element: By.id: lst-ib])
10:56:59.374 INFO - Done: [find element: By.id: lst-ib]
10:56:59.384 INFO - Executing: [send keys: 0 [[FirefoxDriver: firefox on ANY (d591a483-d78f-4269-be1d-1ec2db78d495)] -> id: lst-ib], [s, e, l, e, n, i, u, m]])
10:56:59.748 INFO - Done: [send keys: 0 [[FirefoxDriver: firefox on ANY (d591a483-d78f-4269-be1d-1ec2db78d495)] -> id: lst-ib], [s, e, l, e, n, i, u, m]]
10:56:59.791 INFO - Executing: [send keys to active: []])
10:56:59.809 WARN - Exception thrown
org.openqa.selenium.UnsupportedCommandException: sendKeysToActiveElement
Build info: version: '3.4.0', revision: 'unknown', time: 'unknown'
System info: host: 'ecDev14', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-78-generic', java.version: '1.8.0_144'
Driver info: driver.version: RemoteWebDriver
	at org.openqa.selenium.remote.http.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:220)
	at org.openqa.selenium.remote.http.AbstractHttpCommandCodec.encode(AbstractHttpCommandCodec.java:118)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:158)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:82)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:637)
	at org.openqa.selenium.remote.RemoteExecuteMethod.execute(RemoteExecuteMethod.java:35)
	at org.openqa.selenium.remote.RemoteKeyboard.sendKeys(RemoteKeyboard.java:35)
	at org.openqa.selenium.support.events.internal.EventFiringKeyboard.sendKeys(EventFiringKeyboard.java:42)
	at org.openqa.selenium.remote.server.handler.interactions.SendKeyToActiveElement.call(SendKeyToActiveElement.java:56)
	at org.openqa.selenium.remote.server.handler.interactions.SendKeyToActiveElement.call(SendKeyToActiveElement.java:31)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.openqa.selenium.remote.server.DefaultSession$1.run(DefaultSession.java:176)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
10:56:59.814 WARN - Exception: sendKeysToActiveElement
Build info: version: '3.4.0', revision: 'unknown', time: 'unknown'
System info: host: 'ecDev14', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-78-generic', java.version: '1.8.0_144'
Driver info: driver.version: RemoteWebDriver
10:56:59.838 INFO - Executing: [delete session: 314712ad-98a5-42e8-982d-046a9c1e3f0c])
1503478619892	Marionette	INFO	New connections will no longer be accepted
JavaScript error: resource://shield-recipe-client/lib/RecipeRunner.jsm, line 249: TypeError: log is undefined
10:57:00.431 INFO - Done: [delete session: 314712ad-98a5-42e8-982d-046a9c1e3f0c]
```

**Comment by barancev** [comment]
Oh, I'm sorry I did not mention you're using PHP binding.

New Firefox driver (aka geckodriver) implements new W3C compatible protocol that is quite different from the old one in Actions part: https://w3c.github.io/webdriver/webdriver-spec.html

You have to report the issue to the PHP binding to implement W3C conformant Actions.

**Comment by Alex9179** [comment]
Was this ever resolved? I've just come across this issue with:

- Firefox v57.0.4 (With Gecko)
- Java Server v3.8.1
- phpunit/phpunit-selenium 4.x
- PHP7.2
- phpunit/phpunit: "6.x"

And can't see a way forward with it?

**Comment by jkar9328** [comment]
Having same issue with selenium 3.8 geckodriver 0.20.1

**Comment by GREENpoint** [comment]
Have same issue

**Comment by kkashyap1707** [comment]
Still facing the issue with Firefox 
I/status - selenium standalone version available: 3.14.0 [last]
I/status - chromedriver version available: 2.42 [last]
geckodriver version available: v0.23.0 [last]

Please share the solutions and reopened the issue.


Failed: sendKeysToActiveElement
                Build info: version: '3.14.0', revision: 'aacccce0', time: '2018-08-02T20:13:22.693Z'
                System info: host: 'keshav', ip: '127.0.0.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-137-generic', java.version: '9.0.4'
                Driver info: driver.version: unknown
UnsupportedOperationError: sendKeysToActiveElement
                Build info: version: '3.14.0', revision: 'aacccce0', time: '2018-08-02T20:13:22.693Z'
                System info: host: 'keshav', ip: '127.0.0.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4.4.0-137-generic', java.version: '9.0.4'
                Driver info: driver.version: unknown
                    at Object.checkLegacyResponse (/home/sahil/WebstormProjects/QA_Automation/node_modules/selenium-webdriver/lib/error.js:546:15)
                    at parseHttpResponse (/home/sahil/WebstormProjects/QA_Automation/node_modules/selenium-webdriver/lib/http.js:509:13)
                    at doSend.then.response (/home/sahil/WebstormProjects/QA_Automation/node_modules/selenium-webdriver/lib/http.js:441:30)
                    at process._tickCallback (internal/process/next_tick.js:68:7)
                From: Task: sendKeys
                    at thenableWebDriverProxy.schedule (/home/sahil/WebstormProjects/QA_Automation/node_modules/selenium-webdriver/lib/webdriver.js:807:17)
                    at actions.map.action (/home/sahil/WebstormProjects/QA_Automation/node_modules/selenium-webdriver/lib/actions.js:124:23)
                    at Array.map (<anonymous>)
